pennylane/templates/embeddings/qaoa.py

```diff
@@ -27,59 +27,66 @@
 )
 
 
+@template
+def zz(weight, wires):
+    """Template for decomposition of ZZ coupling.
+
+    Args:
+        wires (list[int]): qubit indices that the template acts on
+    """
+    CNOT(wires=wires)
+    RZ(2 * weight, wires=wires[0])
+    CNOT(wires=wires)
+
+
 def qaoa_feature_encoding_hamiltonian(features, wires):
     """Implements the encoding Hamiltonian of the QAOA embedding.
 
     Args:
         features (array): array of features to encode
-        wires (Sequence[int] or int): `n` qubit indices that the template acts on
+        wires (list[int]): qubit indices that the template acts on
     """
 
-    feature_encoding_wires = wires[: len(features)]
-    remaining_wires = wires[len(features) :]
+    feature_encoding_wires = wires[:len(features)]
+    remaining_wires = wires[len(features):]
 
     broadcast(block=RX, pattern="single", wires=feature_encoding_wires, parameters=features)
     broadcast(block=Hadamard, pattern="single", wires=remaining_wires)
 
 
-def qaoa_ising_hamiltonian(weights, wires, local_fields, l):
+def qaoa_ising_hamiltonian(weights, wires, local_fields):
     """Implements the Ising-like Hamiltonian of the QAOA embedding.
 
     Args:
-        weights (array): array of weights
-        wires (Sequence[int] or int): `n` qubit indices that the template acts on
+        weights (array): array of weights for one layer
+        wires (list[int]): qubit indices that the template acts on
         local_fields (str): gate implementing the local field
-        l (int): layer index
     """
-    # trainable "Ising" ansatz
+
     if len(wires) == 1:
-        local_fields(weights[l][0], wires=wires[0])
+        broadcast(block=local_fields, pattern="single", wires=wires, parameters=weights)
 
     elif len(wires) == 2:
-        # ZZ coupling
-        CNOT(wires=[wires[0], wires[1]])
-        RZ(2 * weights[l][0], wires=wires[0])
-        CNOT(wires=[wires[0], wires[1]])
+        # exception to optimize code: template for 2 wires does not have periodic boundary conditions
+        # since it would repeat the ZZ gate on the same wires
 
+        weights_zz = weights[:1]
+        weights_fields = weights[1:]
+
+        # zz couplings
+        broadcast(block=zz, pattern="chain", wires=wires, parameters=weights_zz)
         # local fields
-        for i, _ in enumerate(wires):
-            local_fields(weights[l][i + 1], wires=wires[i])
+        broadcast(block=local_fields, pattern="single", wires=wires, parameters=weights_fields)
```

Comment:
```suggestion
        broadcast(block=local_fields, pattern="chain", wires=wires, parameters=weights_fields)
```

Reply:
Let me know if I'm missing something with these two suggestions

Reply:
The local fields are broadcasted with a single pattern indeed. (They are the single-spin fields in an Ising model).

Reply:
Oh, this is my bad entirely, thank you so much! :blush: